dhcpv6/async.go

```
@@ -0,0 +1,214 @@
+
+// NewClientAsync creates an asynchronous client
+func NewClientAsync() *ClientAsync {
+	return &ClientAsync{
+		ReadTimeout:  DefaultReadTimeout,
+		WriteTimeout: DefaultWriteTimeout,
+	}
+}
+
+// OpenForInterface starts the client on the specified interface
+func (c *ClientAsync) OpenForInterface(ifname string, bufferSize int) error {
+	addr, err := GetLinkLocalAddr(ifname)
+	if err != nil {
+		return err
+	}
+	c.LocalAddr = &net.UDPAddr{IP: *addr, Port: DefaultClientPort, Zone: ifname}
+	return c.Open(bufferSize)
+}
+
+// Open starts the client
+func (c *ClientAsync) Open(bufferSize int) error {
+	var (
+		addr *net.UDPAddr
+		ok   bool
+		err  error
+	)
+
+	if addr, ok = c.LocalAddr.(*net.UDPAddr); !ok {
+		return fmt.Errorf("Invalid local address: %v not a net.UDPAddr", c.LocalAddr)
+	}
+
+	// prepare the socket to listen on for replies
+	c.connection, err = net.ListenUDP("udp6", addr)
+	if err != nil {
+		return err
+	}
+	c.stopping = make(chan bool)
+	c.sendQueue = make(chan DHCPv6, bufferSize)
+	c.receiveQueue = make(chan DHCPv6, bufferSize)
+	c.packets = make(map[uint32](chan Response))
+	c.packetsLock = sync.Mutex{}
+
+	var ctx context.Context
+	ctx, c.cancel = context.WithCancel(context.Background())
+	go c.listen(ctx)
+	go c.broadcast(ctx)
+
+	return nil
+}
+
+// Close stops the client
+func (c *ClientAsync) Close() {
+	c.cancel()
+	<-c.stopping
+	<-c.stopping
+	close(c.sendQueue)
+	close(c.receiveQueue)
+	c.connection.Close()
+}
+
+func (c *ClientAsync) listen(ctx context.Context) {
+	defer func() { c.stopping <- true }()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case packet := <-c.receiveQueue:
+			c.receive(packet)
+		}
+	}
+}
+
+func (c *ClientAsync) broadcast(ctx context.Context) {
+	defer func() { c.stopping <- true }()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case packet := <-c.sendQueue:
+			c.send(packet)
+		}
+	}
+}
+
+func (c *ClientAsync) send(packet DHCPv6) {
+	transactionID, _ := GetTransactionID(packet)
```

don't ignore the error, log it and return instead. The side effect of ignoring the error is that you will use 0 as transaction ID in the rest of the function

---

This error cannot occur here, we have checked for this packet having a transaction id before puting it to the queue `async.go:191`

---

you can still defensively log it, as you don't know how the code will evolve in the future - maybe with `log.Printf("Warning: this should never happen, there is no transaction ID")`